core/any: remove Provider trait, rename Demand to Request

waynr opened the pull request:
This touches on two WIP features:

* `error_generic_member_access`
  * tracking issue: https://github.com/rust-lang/rust/issues/99301
  * RFC (WIP): https://github.com/rust-lang/rfcs/pull/2895
* `provide_any`
  * tracking issue: https://github.com/rust-lang/rust/issues/96024
  * RFC: https://github.com/rust-lang/rfcs/pull/3192

The changes in this PR are intended to address libs meeting feedback summarized by @Amanieu in https://github.com/rust-lang/rust/issues/96024#issuecomment-1554773172

The specific items this PR addresses so far are:

> We feel that the names "demand" and "request" are somewhat synonymous and would like only one of those to be used for better consistency.

I went with `Request` here since it sounds nicer, but I'm mildly concerned that at first glance it could be confused with the use of the word in networking context.

> The Provider trait should be deleted and its functionality should be merged into Error. We are happy to only provide an API that is only usable with Error. If there is demand for other uses then this can be provided through an external crate.

The net impact this PR has is that examples which previously looked like
```
    core::any::request_ref::<String>(&err).unwramp()
```

now look like
```
    (&err as &dyn core::error::Error).request_value::<String>().unwrap()
```

These are methods that based on the type hint when called return an `Option<T>` of that type. I'll admit I don't fully understand how that's done, but it involves `core::any::tags::Type` and `core::any::TaggedOption`, neither of which are exposed in the public API, to construct a `Request` which is then passed to the `Error.provide` method.

Something that I'm curious about is whether or not they are essential to the use of `Request` types (prior to this PR referred to as `Demand`) and if so does the fact that they are kept private imply that `Request`s are only meant to be constructed privately within the standard library? That's what it looks like to me.

These methods ultimately call into code that looks like:
```
/// Request a specific value by tag from the `Error`.
fn request_by_type_tag<'a, I>(err: &'a (impl Error + ?Sized)) -> Option<I::Reified>
where
    I: tags::Type<'a>,
{
    let mut tagged = core::any::TaggedOption::<'a, I>(None);
    err.provide(tagged.as_request());
    tagged.0
}
```

As far as the `Request` API is concerned, one suggestion I would like to make is that the previous example should look more like this:
```
/// Request a specific value by tag from the `Error`.
fn request_by_type_tag<'a, I>(err: &'a (impl Error + ?Sized)) -> Option<I::Reified>
where
    I: tags::Type<'a>,
{
    let tagged_request = core::any::Request<I>::new_tagged();
    err.provide(tagged_request);
    tagged.0
}
```
This makes it possible for anyone to construct a `Request` for use in their own projects without exposing an implementation detail like `TaggedOption` in the API surface.

Otherwise noteworthy is that I had to add `pub(crate)` on both `core::any::TaggedOption` and `core::any::tags` since `Request`s now need to be constructed in the `core::error` module. I considered moving `TaggedOption` into the `core::error` module but again I figured it's an implementation detail of `Request` and belongs closer to that.

At the time I am opening this PR, I have not yet looked into the following bit of feedback:

> We took a look at the generated code and found that LLVM is unable to optimize multiple .provide_* calls into a switch table because each call fetches the type id from Erased::type_id separately each time and the compiler doesn't know that these calls all return the same value. This should be fixed.

This is what I'll focus on next while waiting for feedback on the progress so far. I suspect that learning more about the type IDs will help me understand the need for `TaggedOption` a little better.

waynr committed:
core/any: remove Provider trait

* remove `impl Provider for Error`
* rename `Demand` to `Request`
* update docstrings to focus on the conceptual API provided by `Request`
* move `core::any::{request_ref, request_value}` functions into `core::error`
* move `core::any::tag`, `core::any::Request`, an `core::any::TaggedOption` into `core::error`
* replace `provide_any` feature name w/ `error_generic_member_access`
* move `core::error::request_{ref,value} tests into core::tests::error module
* update unit and doc tests

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5490929363/jobs/10006959771) [(plain)](https://github.com/rust-lang/rust/commit/649f7dbf1853908bf1aa4527a121eb4721689e1a/checks/14871529348/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[group]Run git config --global core.autocrlf false
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
  lfs: false
  submodules: false
  set-safe-directory: true
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
##[group]Run src/ci/scripts/should-skip-this.sh
src/ci/scripts/should-skip-this.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
##[group]Run src/ci/scripts/verify-channel.sh
src/ci/scripts/verify-channel.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/collect-cpu-stats.sh
src/ci/scripts/collect-cpu-stats.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
##[group]Run src/ci/scripts/install-sccache.sh
src/ci/scripts/install-sccache.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/select-xcode.sh
src/ci/scripts/select-xcode.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/install-clang.sh
src/ci/scripts/install-clang.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/install-wix.sh
src/ci/scripts/install-wix.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/disable-git-crlf-conversion.sh
src/ci/scripts/disable-git-crlf-conversion.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
##[group]Run src/ci/scripts/install-msys2.sh
src/ci/scripts/install-msys2.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/install-mingw.sh
src/ci/scripts/install-mingw.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/install-ninja.sh
src/ci/scripts/install-ninja.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/enable-docker-ipv6.sh
src/ci/scripts/enable-docker-ipv6.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
##[group]Run src/ci/scripts/disable-git-crlf-conversion.sh
src/ci/scripts/disable-git-crlf-conversion.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/verify-line-endings.sh
src/ci/scripts/verify-line-endings.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/verify-backported-commits.sh
src/ci/scripts/verify-backported-commits.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
##[group]Run src/ci/scripts/verify-stable-version-number.sh
src/ci/scripts/verify-stable-version-number.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
##[group]Run src/ci/scripts/run-build-from-ci.sh
src/ci/scripts/run-build-from-ci.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: 649f7dbf1853908bf1aa4527a121eb4721689e1a
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error: unclosed HTML tag `T`
   --> library/core/src/any.rs:114:67
    |
114 | //! `request_value` are available to simplify obtaining an `Option<T>` for a given type.
    |
    = note: `-D rustdoc::invalid-html-tags` implied by `-D warnings`
help: try marking as source code
    |
    |
114 | //! `request_value` are available to simplify obtaining an ``Option<T>`` for a given type.

error: could not document `core`

Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2021 --crate-type lib --crate-name core library/core/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/doc -Zunstable-options --check-cfg 'names()' --check-cfg 'values()' --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat -C metadata=6f12f42ca034c041 -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/release/deps --cfg=bootstrap -Csymbol-mangling-version=legacy -Zunstable-options -Zunstable-options '--check-cfg=values(bootstrap)' '--check-cfg=values(stdarch_intel_sde)' '--check-cfg=values(no_fp_fmt_parse)' '--check-cfg=values(no_global_oom_handling)' '--check-cfg=values(no_rc)' '--check-cfg=values(no_sync)' '--check-cfg=values(freebsd12)' '--check-cfg=values(freebsd13)' '--check-cfg=values(backtrace_in_libstd)' '--check-cfg=values(target_env,"libnx")' '--check-cfg=values(target_arch,"asmjs","spirv","nvptx","xtensa")' --document-private-items --document-hidden-items -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.72.0-nightly (c9a81b879 2023-07-07)' '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")' '-Zcrate-attr=warn(rust_2018_idioms)' -Z unstable-options --resource-suffix 1.72.0 --markdown-css rust.css --markdown-no-toc --index-page /checkout/src/doc/index.md` (exit status: 1)
Build completed unsuccessfully in 0:00:08
```

</details>

waynr commented:
One problem I'm currently running into when attempting to reproduce the failed test by running `x doc` is that there are rust tools (`rustfmt` and `rust-analyzer` that depend on `thiserror` which in turn actually depends on the unstable features my PR deals with):

```
Building compiler artifacts (stage0 -> stage1, x86_64-unknown-linux-gnu)
   Compiling cfg-if v1.0.0
   Compiling stable_deref_trait v1.2.0
   Compiling once_cell v1.18.0
   Compiling smallvec v1.10.0
   Compiling scopeguard v1.1.0
   Compiling log v0.4.19
   Compiling pin-project-lite v0.2.10
   Compiling libc v0.2.147
   Compiling memchr v2.5.0
   Compiling rustc-hash v1.1.0
   Compiling bitflags v1.3.2
   Compiling typenum v1.16.0
   Compiling zerofrom v0.1.2
   Compiling equivalent v1.0.0
   Compiling hashbrown v0.14.0
   Compiling thin-vec v0.2.12
   Compiling instant v0.1.12
   Compiling lock_api v0.4.10
   Compiling linux-raw-sys v0.3.8
   Compiling psm v0.1.21
   Compiling fastrand v1.9.0
   Compiling arrayvec v0.7.4
   Compiling either v1.8.1
   Compiling elsa v1.7.1
   Compiling yoke v0.7.1
   Compiling rustc_graphviz v0.0.0 (/home/wayne/projects/rust-lang/rust/compiler/rustc_graphviz)
   Compiling tracing-core v0.1.30
   Compiling itertools v0.10.5
   Compiling rustc_arena v0.0.0 (/home/wayne/projects/rust-lang/rust/compiler/rustc_arena)
   Compiling ena v0.14.2
   Compiling ppv-lite86 v0.2.17
   Compiling thiserror v1.0.40
   Compiling zerovec v0.9.4
error[E0432]: unresolved imports `std::any::Demand`, `std::any::Provider`
 --> /home/wayne/.cargo/registry/src/index.crates.io-6f17d22bba15001f/thiserror-1.0.40/src/provide.rs:1:16
  |
1 | use std::any::{Demand, Provider};
  |                ^^^^^^  ^^^^^^^^ no `Provider` in `any`
  |                |
  |                no `Demand` in `any`

For more information about this error, try `rustc --explain E0432`.
error: could not compile `thiserror` (lib) due to previous error
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:01:39
```

Does this mean I'll have to update `thiserror` to reflect the new API changes and include updating that to a new version in this PR? I was already considering that since the behavior I want is to enable automatic backtrace capture by `thiserror`.

cuviper commented:
This is a big change that I don't have any context on, so I'm reassigning based on the prior feedback.

r? @Amanieu

thomcc commented:
Can you write up new API somewhere (or at least the changes)? It's hard to get the full picture from the diff.

the8472 commented:
> One problem I'm currently running into when attempting to reproduce the failed test by running `x doc` is that there are rust tools (`rustfmt` and `rust-analyzer` that depend on `thiserror` which in turn actually depends on the unstable features my PR deals with):

Are the nightly features of those crates critical or can we disable them?

waynr commented:
@thomcc Okay here's what I understand of the full API changes for `provide_any` and `error_generic_member_access` prior to this PR:

* new public trait `core::any::Provider` with one method, `fn provide<'a>(&'a self, demand: &mut Demand<'a>)`
* new public type `core::any::Demand` with several methods related to checking for or retrieving values or references:
  * `pub fn provide_value<T>(&mut self, value: T) -> &mut Self`
  * `pub fn provide_value_with<T>(&mut self, fulfil: impl FnOnce() -> T) -> &mut Self`
  * `pub fn provide_ref<T: ?Sized + 'static>(&mut self, value: &'a T) -> &mut Self`
  * `pub fn provide_ref_with<T: ?Sized + 'static>( &mut self, fulfil: impl FnOnce() -> &'a T,) -> &mut Self`
  * `pub fn would_be_satisfied_by_value_of<T>(&self) -> bool`
  * `pub fn would_be_satisfied_by_ref_of<T>(&self) -> bool`
* new functions for working with Provider and Demand:
  * `pub fn request_value<'a, T>(provider: &'a (impl Provider + ?Sized)) -> Option<T>`
  * `pub fn request_ref<'a, T>(provider: &'a (impl Provider + ?Sized)) -> Option<&'a T>`
* new public method with default noop implementation for trait `core::error::Error`: `fn provide<'a>(&'a self, demand: &mut Demand<'a>) {}`
* `Provider` implementation for `dyn Error`

Again, this is the state of the `provide_any` and `error_generic_member_access` prior to this PR. The changes this PR makes is:

* remove the new public trait `core::any::Provider`
* rename `core::any::Demand` to`core::any::Request`
* remove `core::any::request_ref` and `core::any::request_value` functions
  * but it adds analogous `request_ref` and `request_value` methods to `<dyn Error>` 
    * this is necessary because the current `Request` API doesn't allow instances to be constructed; `request_ref` and `request_value` internally do the work of constructing the `Request` instance and returning an `Option<T>` where T is the type annotation provided when calling `request_value` or `request_ref`.

Not an API change, but this does also enable `error_in_core`, `error_generic_member_access` in `library/core/tests/lib.rs` which if I understand correctly is necessary because tests in `library/core/tests/any.rs` now need to import `core::error::Error` for some of their test cases (since `Error` is the only type that makes use of the new `Request` type being tested).

Also worth noting is that I had to add `trait_upcasting` which I don't fully understand but I think is because of test cases that look like this:

```
@@ -188,18 +196,18 @@ fn test_provider_boxed() {
 fn test_provider_concrete() {
     let obj = SomeConcreteType { some_string: "hello".to_owned() };

-    assert_eq!(&**request_ref::<String>(&obj).unwrap(), "hello");
-    assert_eq!(&*request_value::<String>(&obj).unwrap(), "bye");
-    assert_eq!(request_value::<u8>(&obj), None);
+    assert_eq!(&**(&obj as &dyn std::error::Error).request_ref::<String>().unwrap(), "hello");
+    assert_eq!(&*(&obj as &dyn std::error::Error).request_value::<String>().unwrap(), "bye");
+    assert_eq!((&obj as &dyn std::error::Error).request_value::<u8>(), None);
 }

-trait OtherTrait: Provider {}
+trait OtherTrait: std::error::Error {}

 impl OtherTrait for SomeConcreteType {}

 impl dyn OtherTrait {
     fn get_ref<T: 'static + ?Sized>(&self) -> Option<&T> {
-        request_ref::<T>(self)
+        <dyn core::error::Error>::request_ref_from::<T>(self)
     }
 }
```

Where `OtherTrait` requires that any implementing types also implement the dependency trait, in this case `trait OtherTrait: std::error::Error`. What I don't understand is why that feature wasn't required to be enabled prior to this PR since something similar was going on with `OtherTrait: Provider` :thinking:

waynr commented:
> @thomcc Okay here's what I understand of the full API changes for provide_any and error_generic_member_access prior to this PR

To be absolutely clear, when I say "prior to this PR" I mean that these are unstable features that someone else worked on which are already in master branch. I am coming to these features late in the game picking up where other people left off and attempting to help address feedback in the tracking issues for these features.

waynr commented:
> > One problem I'm currently running into when attempting to reproduce the failed test by running `x doc` is that there are rust tools (`rustfmt` and `rust-analyzer` that depend on `thiserror` which in turn actually depends on the unstable features my PR deals with):
> 
> Are the nightly features of those crates critical or can we disable them?

By "those crates" are you referring to the `rustfmt` and `rust-analyzer` crates? If so, I'm not sure what I would need to do to disable `provide_any` or `error_generic_member_access` for either of them. I tried this:

```
commit bd3326d2276a490e6b212e9816f998cc76ecb6dd
Merge: 0c981cdd9cf 45dd2c0a6aa
Author: wayne warren <wayne.warren.s@gmail.com>
Date:   Sat Jul 8 12:11:33 2023 -0600

    WIP on master: 0c981cdd9cf core/any: remove references to Provider, update tests

diff --cc src/tools/rust-analyzer/crates/rust-analyzer/Cargo.toml
index 5b72d57560b,5b72d57560b..6a0643da4dc
--- a/src/tools/rust-analyzer/crates/rust-analyzer/Cargo.toml
+++ b/src/tools/rust-analyzer/crates/rust-analyzer/Cargo.toml
@@@ -51,7 -51,7 +51,7 @@@ always-assert = "0.1.2
  # These dependencies are unused, but we pin them to a version here to restrict them for our transitive dependencies
  # so that we don't pull in duplicates of their dependencies like windows-sys and syn 1 vs 2
  # these would pull in serde 2
--thiserror = "=1.0.39"
++thiserror = { version = "=1.0.39", features = [] }
  serde_repr = "=0.1.11"
  # these would pull in windows-sys 0.45.0
  mio = "=0.8.5"
diff --cc src/tools/rustfmt/Cargo.toml
index 8c312f47a28,8c312f47a28..a16accc33d7
--- a/src/tools/rustfmt/Cargo.toml
+++ b/src/tools/rustfmt/Cargo.toml
@@@ -50,7 -50,7 +50,7 @@@ regex = "1.5
  serde = { version = "1.0.160", features = ["derive"] }
  serde_json = "1.0"
  term = "0.7"
--thiserror = "1.0.40"
++thiserror = { version = "1.0.40", features = [] }
  toml = "0.7.4"
  unicode-segmentation = "1.9"
  unicode-width = "0.1"
```

But end up seeing the same error anyway when I run `x doc`. @the8472 where else might I look to disable them?

waynr commented:
> Are the nightly features of those crates critical or can we disable them?

So my vague understanding is that `thiserror` bases its usage of the `provide_any` features [on the rust feature flag](https://github.com/dtolnay/thiserror/blob/225adab854715459bc81dd3e788805debcf7f310/src/lib.rs#L239):
```
#![cfg_attr(provide_any, feature(provide_any))]
```

If I understand correctly, then `provide_any` is enabled on the `rustfmt` and `rust-analyzer` crates because when building using `x doc` they are built with nightly features of the standard library enabled and `provide_any` is considered a nightly feature (is that right?).

So to be clear I don't think that `rustfmt` and `rust-analyzer` are using these features intentionally, they just happen to be enabled when compiling `thiserror` on nightly.

I can think of three ways to fix the problem for this PR:

* add thiserror-specific feature flag to thiserror (ie open a PR against that repo) that allows us to explicitly disable usage of `provide_any` features eg
```
#![cfg_attr(all(provide_any, enable_provide_any), feature(provide_any))]
```
* somehow remove `provide_any` from the default set of enabled nightly features
* work on a PR against thiserror in parallel with this PR to keep the changes sync'd

My preference would be the last option. I will test the `thiserror` branch I work on builds using a toolchain that i build from the `rust` repo. For the sake of getting CI to pass on this PR I think I would need to pin that version to a branch in the cargo.toml of `rustfmt` and `rust-analyzer` (I'm still a little fuzzy on how the compilation stages/steps work)

One problem I can anticipate with this approach is the need to work out a merge strategy between rust maintainers and thiserror maintainers since there is kind of a chicken-and-egg problem here:
* would we merge into `thiserror` first, leaving its master/main branch broken until this PR merges into rust master?
  * I would think not because this would break anyone using `thiserror` on nightly.
* or would we merge into `rust` first? but could we if tools aren't building?


One way to resolve the issue might be to merge this PR with a git-pinned version, for which there seems to be precedent:
```
./compiler/rustc_codegen_gcc/Cargo.toml
25:gccjit = { git = "https://github.com/antoyo/gccjit.rs" }
```
Then we would merge the `thiserror` PR, get a release for it, and update the `rustfmt` and `rust-analyzer` crates with the new `thiserror` version. At this point users who haven't updated their old nightly would just need to update it to keep building their project on nightly.

I'm probably overthinking this. If nightly is temporarily broken for a particular crate users can just hold off on updating until there is a good nightly build of the toolchain. 

I hope none of this sounds crazy, I'm just trying to be thorough in my response!

workingjubilee commented:
bikesheddy/nonblocking comment: I agree that `Demand` makes this API look a lot different from "some networking thing", whereas the "nicer" sounding `Request` is indeed a bit confusing, and thus would indeed slightly prefer the meaner type. :smiling_imp:

It wouldn't be the first overloaded name in std, though, if the consensus was otherwise, so it's not the end of the world either way.

waynr commented:
> bikesheddy/nonblocking comment: I agree that Demand makes this API look a lot different from "some networking thing", whereas the "nicer" sounding Request is indeed a bit confusing, and thus would indeed slightly prefer the meaner type. 

Since we're getting bike-sheddy...

In addition to the lack of manners, I also don't like `Demand` because the word implies a sense of intolerance for not getting what is being demanded whereas the value/reference being returned here is wrapped in an `Option<T>` which implies the expectation that it may not be fulfilled. Therefore, for `Demand` to be semantically correct shouldn't we return a `Result`?  

So instead of `Demand` or `Request`, what do you think about `Query`? I feel like this is better-aligned with returning an `Option<T>` than `Demand` and possibly even better than `Request`.

workingjubilee commented:
clearly, if we want to be that polite, we should make it an Invitation! /silly

waynr commented:
@Amanieu gentle ping for review, or suggestions as to who might have bandwidth to review in the near future. I'll also ping the `t-release/triage` channel in zulip as suggested in the [dev guide](https://rustc-dev-guide.rust-lang.org/contributing.html#waiting-for-reviews).

@thomcc did [my comment](https://github.com/rust-lang/rust/pull/113464#issuecomment-1627456712) satisfy your request for a summary of changes?

@the8472 regarding my previous report of `x doc` failing -- I can no longer reproduce that issue locally since I removed the `provide_any` flag in my latest commit but did see [a suggestion from dtolany](https://github.com/dtolnay/thiserror/pull/243#pullrequestreview-1553150076) that it may have been some kind of bug in the bootstrap tooling related to unexpected standard lib used for build scripts:

> That error is caused by a bug in rustc's bootstrap where, in some stages, the standard library that gets picked up if $RUSTC is run from a build script is not the standard library that the rest of the crate will get built against.

I don't fully understand this yet. It seems strange to me that it was happening at all given that the `thiserror` build script performs a trybuild-based probe to detect the availability of the `Provider` feature: https://github.com/dtolnay/thiserror/blob/543e1237d600949398a7c2e301722ce63cabbd95/build.rs#L22

I guess if the nightly toolchain automatically enables `provide_any` than the build script is kind of a noop?

Anyway, to answer your original question again more concisely:
> Are the nightly features of those crates critical or can we disable them?

They are effectively disabled with the latest changes in this PR.

Amanieu commented on library/core/src/any.rs:
```diff
@@ -83,72 +83,115 @@
+//!     fn provide<'a>(&'a self, request: &mut Request<'a>) {
+//!         request
+//!             .provide_ref::<MyBacktrace>(&self.backtrace)
+//!             .provide_ref::<dyn std::error::Error + 'static>(&self.source);
```
Is this a pattern that we want to encourage? The source should be exposed via the `source` trait method, not `provide`.

waynr replied:
Hmm, good point. For what it's worth I've mostly been naively copying and pasting stuff here and there with modifications to support removing the Provider trait. Will remove this and other references to source in the examples for `Request`.

Amanieu commented on library/core/src/any.rs:
```diff
@@ -1193,23 +1208,24 @@ impl<'a> Demand<'a> {
 ///////////////////////////////////////////////////////////////////////////////
 
-mod tags {
+#[unstable(feature = "error_generic_member_access", issue = "99301")]
```
These feature tags should not be needed for things that are `pub(crate)`.

waynr replied:
d'oh!

Amanieu commented on library/core/src/error.rs:
````diff
@@ -412,6 +412,60 @@ impl dyn Error {
+/// fn get_string(err: &impl Error) -> String {
+///     (err as &dyn Error).request_value::<String>().unwrap()
+/// }
+/// ```
````
This doc comment should be moved into the `impl dyn Error` version of this function that is actually `pub`.

Amanieu commented on library/core/src/error.rs:
```diff
@@ -219,13 +214,18 @@ impl<'a> dyn Error + 'a {
     /// Request a reference of type `T` as context about this error.
     #[unstable(feature = "error_generic_member_access", issue = "99301")]
     pub fn request_ref<T: ?Sized + 'static>(&'a self) -> Option<&'a T> {
-        core::any::request_ref(self)
+        request_ref(self)
     }
 
     /// Request a value of type `T` as context about this error.
     #[unstable(feature = "error_generic_member_access", issue = "99301")]
     pub fn request_value<T: 'static>(&'a self) -> Option<T> {
-        core::any::request_value(self)
+        request_value(self)
+    }
```
Could you add these to the `Error` trait with a `where Self: Sized` bound, so they can also be used with normal error types that are not `dyn Error`. This is useful since there is no other way to create a `Request`.

waynr replied:
Do you mean we should keep the `dyn Error` impls and simply add new `Error` trait methods with default function bodies? Or should we discard the `dyn Error` impls? If `request_value` and `request_ref` are trait methods then they automatically become available for `dyn Error`s, right?

waynr replied:
Oh I think I may have realized the answer to this question -- we need both because the `where Self: Sized` bound doesn't apply to `dyn Error`, right?

waynr replied:
Okay I think my last comment is essentially correct because when I remove all `request_ref` and `request_value` from the variations on `dyn Error` I get
```
   Compiling std v0.0.0 (/home/wayne/projects/rust-lang/rust/library/std)
error: the `request_ref` method cannot be invoked on a trait object
   --> library/std/src/error.rs:460:53
    |
460 |         let backtrace = (&self.error as &dyn Error).request_ref();
    |                                                     ^^^^^^^^^^^
    |
   ::: /home/wayne/projects/rust-lang/rust/library/core/src/error.rs:218:21
    |
218 |         where Self: Sized
    |                     ----- this has a `Sized` requirement

error: the `request_ref` method cannot be invoked on a trait object
   --> library/std/src/error.rs:464:73
    |
464 |                 .map(|source| source.sources().find_map(|source| source.request_ref()))
    |                                                                         ^^^^^^^^^^^
    |
   ::: /home/wayne/projects/rust-lang/rust/library/core/src/error.rs:218:21
    |
218 |         where Self: Sized
    |                     ----- this has a `Sized` requirement

error: could not compile `std` (lib) due to 2 previous errors
```
Line 460 can be fixed by changing `(&self.error as &dyn Error)` to `self.error` but on line 464 the `source` in `source.request_ref()` is the return value of the `Error` trait method `source` which has type `&(dyn Error + 'static)`.

Just to be clear, we do want to support generic member access on `dyn Error`, right? But when I don't remove the `request_value` and request_ref` methods from the `dyn Error` variations I get this kind of error:

```
error[E0034]: multiple applicable items in scope
   --> library/core/src/error.rs:404:22
    |
404 |         <dyn Error>::request_value(self)
    |                      ^^^^^^^^^^^^^ multiple `request_value` found
    |
note: candidate #1 is defined in an impl for the type `(dyn error::Error + 'a)`
   --> library/core/src/error.rs:290:5
    |
290 |     pub fn request_value<T: 'static>(&'a self) -> Option<T> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `(dyn error::Error + Send + 'static)`
   --> library/core/src/error.rs:368:5
    |
368 |     pub fn request_value<T: 'static>(&self) -> Option<T> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #3 is defined in an impl for the type `(dyn error::Error + Send + Sync + 'static)`
   --> library/core/src/error.rs:403:5
    |
403 |     pub fn request_value<T: 'static>(&self) -> Option<T> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #4 is defined in the trait `error::Error`
   --> library/core/src/error.rs:238:5
    |
238 | /     fn request_value<'a, T: 'static>(&'a self) -> Option<T>
239 | |         where Self: Sized
    | |_________________________^
help: disambiguate the method for candidate #4
    |
404 |         <dyn error::Error as error::Error>::request_value(self)
    |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

For line 404 if I try doing something like `self.request_value()`, I get the aforementioned `error: the `request_value` method cannot be invoked on a trait object` error. Definitely open to suggestions here...

Amanieu replied:
Ah, I didn't think they would conflict thanks to the `Self: Sized` bound, but that's probably not possible due to the way name resolution works. In that case it's probably better to keep them as free functions outside the trait, and remove both the ones in the trait (that I mistakenly suggested) and the ones in `impl dyn Error`. The goal here is to have one consistent way to request values that works for both concrete error types and `dyn Error`.

waynr replied:
Oh interesting, so don't even keep the `impl dyn Error` methods? Will do!

waynr replied:
Okay so I tried using free functions and removing the `impl dyn Error` methods as suggested but I am now running into this error:
```
error[E0277]: the size for values of type `dyn core::error::Error` cannot be known at compilation time
   --> library/std/src/error.rs:466:78
    |
466 |                 .map(|source| source.sources().find_map(|source| request_ref(source)))
    |                                                                  ----------- ^^^^^^ doesn't have a size known at compile-time
    |                                                                  |
    |                                                                  required by a bound introduced by this call
    |
    = help: the trait `core::marker::Sized` is not implemented for `dyn core::error::Error`
note: required by a bound in `request_ref`
   --> /home/wayne/projects/rust-lang/rust/library/core/src/error.rs:409:36
    |
409 | pub fn request_ref<'a, T>(err: &'a impl Error) -> Option<&'a T>
    |                                    ^^^^^^^^^^ required by this bound in `request_ref`
```

This after removing the `+ &Sized` bound in that parameter's type:
```
-fn request_value<'a, T>(err: &'a (impl Error + ?Sized)) -> Option<T>
+pub fn request_value<'a, T>(err: &'a impl Error) -> Option<T>
```

The error is literally saying that `impl Error` implies a `core::marker::Sized` bound on the type parameter, but I don't see that mentioned in [the docs](https://doc.rust-lang.org/reference/types/impl-trait.html) on `impl Error` which I understand to be telling me that in parameter position `impl Error` is similar in purpose to a generic type parameter.

My second or third thought about what might be going on here is that there could be a `Self: Sized` bound on a different `Error` trait method, but I don't see that. But then I thought that maybe there is an `impl Error` block somewhere doing that and sure enough I found it: https://github.com/rust-lang/rust/blob/master/library/core/src/error.rs#L437

Could this line be the reason that `impl Error` implies a `Sized` bound?

Another question that's been stewing in my puny brain over the past couple days while pondering this is, what does this mean?:

> The goal here is to have one consistent way to request values that works for both concrete error types and dyn Error

In my mind the prototypical use case of generic member access for `Error` (or for any type for that matter) is that we want to offer the ability for one library that knows nothing about the concrete `Error` types it might be dealing with to access a value that might potentially be provided by some concrete types but not others. To be more specific, I would like to see a "frontend" error-handling library like `anyhow` be able to access data (specifically, backtraces) stored on errors it can't be certain is available because the concrete type of an `Error` that such frontends encounter will never be explicitly known by `anyhow` (my understanding is that `anyhow` would be usually be accessing `dyn Error`).

I'll admit that what I just wrote confuses me a little, but the train of reasoning I'm trying to reach here is -- does it make sense to want to perform generic member access on concrete error types? If you have access to the concrete type, you could just perform field access of the member that you're interested in. But I suspect I am misunderstanding what you meant (hence the word salad here).

Instead of the free function, would it make sense to instead have `request_ref` and `request_value` methods defined on both `dyn Error` and  the aforementioned `impl<'a, T: Error + ?Sized> Error for &'a T`? Kind of getting into "don't know what i'm talking about" territory here in the sense that I'm not sure if this addresses your concern about consistency...

Amanieu replied:
`T: Sized` bounds are always implicitly added unless specifically opted-out-of with `T: ?Sized`. This has nothing to do with `Error` but is instead inherent to the Rust language.

You shouldn't need to change the signature of `request_value`/`request_ref`, it is already correct with `err: &'a (impl Error + ?Sized)`.

waynr replied:
d'oh! Thanks for the quick response.

waynr replied:
I think the reason I removed that was because I was confused about some other compiler errors...gonna take a closer at those now.

waynr replied:
Okay this is the error that may have been confusing me before:
```
error[E0277]: the size for values of type `dyn std::error::Error` cannot be known at compilation time
   --> library/core/tests/any.rs:199:36
    |
199 |     assert_eq!(request_value::<u8>(&obj), None);
    |                ------------------- ^^^^ doesn't have a size known at compile-time
    |                |
    |                required by a bound introduced by this call
    |
    = help: the trait `Sized` is not implemented for `dyn std::error::Error`
    = help: the trait `std::error::Error` is implemented for `Box<T>`
    = note: required for `Box<dyn std::error::Error>` to implement `std::error::Error`
note: required by a bound in `request_value`
   --> /home/wayne/projects/rust-lang/rust/library/core/src/error.rs:387:44
    |
387 | pub fn request_value<'a, T>(err: &'a (impl Error + ?Sized)) -> Option<T>
    |                                            ^^^^^ required by this bound in `request_value`
```

Since I didn't know that `?` meant to opt-out of `T: Sized` bounds I incorrectly thought removing `?Sized` would fix this, but it actually just makes the compile fail in a slightly different way.

The full body of the specific test that's failing is:
```
// Test the provide and request mechanisms with a boxed trait object.
#[test]
fn test_provider_boxed() {
    let obj: Box<dyn std::error::Error> =
        Box::new(SomeConcreteType { some_string: "hello".to_owned() });

    assert_eq!(request_value::<String>(&*obj).unwrap(), "bye");
    assert_eq!(request_value::<u8>(&obj), None);
}
```

The definition of `request_value` used here is:
```
pub fn request_value<'a, T>(err: &'a (impl Error + ?Sized)) -> Option<T>
where
    T: 'static,
{
    request_by_type_tag::<'a, tags::Value<T>>(err)
}
```

I'm still a confused by this, but I'm also getting super tired so I'll try to come back to it in the morning.

Amanieu replied:
It seems that `Box<E>` only implement `Error` when `E: Error + Sized`. See #104485.

You can work around it by using `&*obj` in the test.

Amanieu commented on library/core/src/error.rs:
```diff
@@ -219,13 +214,18 @@ impl<'a> dyn Error + 'a {
+        request_value(self)
+    }
+
+    /// Request a ref of type `T` from the given `Self` instance.
+    pub fn request_ref_from<T: ?Sized + 'static>(this: &'a Self) -> Option<&'a T> {
+        this.request_ref()
```
What is the point of this method?

waynr replied:
Honestly, it's been about a month so I don't remember. My first impulse is to say "it's a naive copy pasta" but I only see it being added in the diff, not removed elsewhere. Maybe I copied a code example from an RFC and wanted to make doctests compile/pass... :thinking:

waynr replied:
Oh yeah, here it is: https://github.com/rust-lang/rfcs/pull/3461/files#diff-bb91f37510dc7a6369183c40dbdcfb52164335efa7a04f005a906d0006754d27R419

waynr replied:
No wait, I'm the one who added that to the RFC: https://github.com/rust-lang/rfcs/pull/3461/commits/0cffa52cca1a4b12d8364b15fb67f561a9e8b8a7#diff-bb91f37510dc7a6369183c40dbdcfb52164335efa7a04f005a906d0006754d27R419

...what in the devil?

waynr replied:
I'll remove it :)

Amanieu commented on library/core/tests/any.rs:
```diff
@@ -166,40 +173,41 @@ impl Provider for SomeConcreteType {
 
-trait OtherTrait: Provider {}
+trait OtherTrait: std::error::Error {}
 
 impl OtherTrait for SomeConcreteType {}
 
 impl dyn OtherTrait {
     fn get_ref<T: 'static + ?Sized>(&self) -> Option<&T> {
-        request_ref::<T>(self)
+        <dyn core::error::Error>::request_ref_from::<T>(self)
     }
 }
```
I would just remove this test, it is not supported with `Error`, only with the generic `Provider` trait which is now being removed.

Amanieu commented on library/core/tests/any.rs:
```diff
@@ -166,40 +173,41 @@ impl Provider for SomeConcreteType {
-    assert_eq!(&**request_ref::<String>(&obj).unwrap(), "hello");
-    assert_eq!(&*request_value::<String>(&obj).unwrap(), "bye");
-    assert_eq!(request_value::<u8>(&obj), None);
+    assert_eq!(&**(&obj as &dyn std::error::Error).request_ref::<String>().unwrap(), "hello");
+    assert_eq!(&*(&obj as &dyn std::error::Error).request_value::<String>().unwrap(), "bye");
+    assert_eq!((&obj as &dyn std::error::Error).request_value::<u8>(), None);
```
Once you add the methods to the error trait as I previously suggested, this can be simplified.

waynr replied:
Even using the free functions that we ultimately landed on simplified this. Thanks for the suggestion!

waynr commented:
@Amanieu I believe I have addressed all your feedback. I also moved some tests from `library/core/tests/any.rs` into a newly-created file `library/core/tests/error.rs` since even though they are making use of `core::any::Request` in some of the test helper code, the tests themselves actually exercise `core::error::request_{ref,value}`. But I added this as the last commit in my series of changes so if that reasoning doesn't jive with you it's easy to revert.

I mentioned this in a comment on #104485, but I am curious what makes the addition of `?Sized` as a bound to generic type parameter a backwards incompatible/breaking change (this claim was made on #104485). Doesn't `?Sized` merely relax/remove the implicit `Sized` bound on type parameters? With the addition of `?Sized` to a type parameter, types that are `Sized` should still be valid if I understand correctly (and I often don't :slightly_smiling_face:). If we could relax that `Sized` requirement on `Box<dyn Error>` then we could keep the test I commented out with a NOTE attempting to explain this situation in my recent updates.

Anyway, thanks again for your patience and quick responses to my questions!

Amanieu commented on library/core/src/error.rs:
```diff
@@ -172,8 +178,8 @@ pub trait Error: Debug + Display {
+    ///     fn provide<'a>(&'a self, request: &mut Request<'a>) {
+    ///         request
     ///             .provide_ref::<MyBacktrace>(&self.backtrace)
     ///             .provide_ref::<dyn std::error::Error + 'static>(&self.source);
```
The source should be provided by the `source` method, not `provide`.

Amanieu commented on library/core/src/any.rs:
```diff
@@ -1220,15 +1223,16 @@ mod tags {
     /// module, although crates may also implement their own tags for more
     /// complex types with internal lifetimes.
-    pub trait Type<'a>: Sized + 'static {
+    #[unstable(feature = "error_generic_member_access", issue = "99301")]
```
This is not a public API and shouldn't have a stability attribute.

waynr replied:
Without it I get the following compile error:

```
error: trait has missing stability attribute
    --> library/core/src/any.rs:1226:5
     |
1226 | /     pub(crate) trait Type<'a>: Sized + 'static {
1227 | |         /// The type of values which may be tagged by this tag for the given
1228 | |         /// lifetime.
1229 | |         type Reified: 'a;
1230 | |     }
     | |_____^

error: could not compile `core` (lib) due to previous error
```

Amanieu replied:
I don't think this trait needs to be `pub(crate)` since all of the users are in this module. If that still doesn't work then  mark it with a comment so that it doesn't accidentally get stabilized along with the feature.

waynr replied:
There are several places outside the `tags` module but inside the `error` module where it is referenced so I'll leave the comment.

waynr replied:
Okay now that it's moved into the `error` module I _don't_ get that "trait has missing stability attribute" error. I hadn't tried removing it after moving the `tags` module and `Request` type into the `error` module. I'll try removing the stability attribute from other items in the `tags` module also.

Amanieu commented on library/core/src/any.rs:
```diff
@@ -1290,7 +1294,7 @@ unsafe impl<'a, I: tags::Type<'a>> Erased<'a> for TaggedOption<'a, I> {
 }
 
-#[unstable(feature = "provide_any", issue = "96024")]
+#[unstable(feature = "error_generic_member_access", issue = "99301")]
```
This is not a public API and shouldn't have a stability attribute.

Amanieu commented on library/core/src/any.rs:
```diff
@@ -803,111 +831,23 @@ pub const fn type_name_of_val<T: ?Sized>(_val: &T) -> &'static str {
 
 ///////////////////////////////////////////////////////////////////////////////
-// Provider trait
+// Request and its methods
```
All this still should be moved out of `any` and into `error`.

waynr replied:
I wondered about that whenever I first started working on this PR. My reasoning for leaving it here was that it seems more semantically aligned with the content of `any` than `error` and even though we are currently rejecting the `Provider` trait for the sake of moving forward with `error_generic_member_access`, it seemed to me possible that one of the following might happen in the future:

* other standard library types than `Error` might end up using it
* the consensus around the `Provider` trait itself (or something similar) could change

But I'll go ahead and move it assuming you've already thought about these things. Worst case scenario, I suppose the implementation could be moved back into `any` and `error` could continue to provide access to it in its API surface for backwards compatibility's sake.

waynr commented:
@Amanieu I've just push a few commits to address your latest round of feedback. Thanks!

Amanieu commented:
LGTM! Can you squash the commit and then I will merge it.

Amanieu commented:
@bors r+

bors commented:
:hourglass: Testing commit a646b3996510fb79a8fd4701f888542e9b280504 with merge 4cea2bc339db4efdc0fbcf44760c2e9a75d73e31...
<!-- homu: {"type":"BuildStarted","head_sha":"a646b3996510fb79a8fd4701f888542e9b280504","merge_sha":"4cea2bc339db4efdc0fbcf44760c2e9a75d73e31"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5859193336/job/15887444497)
Approved by: Amanieu
Pushing 4cea2bc339db4efdc0fbcf44760c2e9a75d73e31 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Amanieu","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5859193336/job/15887444497"},"merge_sha":"4cea2bc339db4efdc0fbcf44760c2e9a75d73e31"} -->

rust-timer commented:
Finished benchmarking commit (4cea2bc339db4efdc0fbcf44760c2e9a75d73e31): [comparison URL](https://perf.rust-lang.org/compare.html?start=475be26d99b8e3898c482e0ac42ab7bce3d511bf&end=4cea2bc339db4efdc0fbcf44760c2e9a75d73e31&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=475be26d99b8e3898c482e0ac42ab7bce3d511bf&end=4cea2bc339db4efdc0fbcf44760c2e9a75d73e31&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=475be26d99b8e3898c482e0ac42ab7bce3d511bf&end=4cea2bc339db4efdc0fbcf44760c2e9a75d73e31&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.5% | [-1.5%, -1.5%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=475be26d99b8e3898c482e0ac42ab7bce3d511bf&end=4cea2bc339db4efdc0fbcf44760c2e9a75d73e31&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=475be26d99b8e3898c482e0ac42ab7bce3d511bf&end=4cea2bc339db4efdc0fbcf44760c2e9a75d73e31&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 632.631s -> 632.353s (-0.04%)
**Artifact size**: 346.50 MiB -> 346.36 MiB (-0.04%)

RalfJung commented on library/core/src/error.rs:
```diff
@@ -123,16 +123,20 @@ pub trait Error: Debug + Display {
     /// #![feature(error_generic_member_access)]
     /// use core::fmt;
-    /// use core::any::Demand;
+    /// use core::error::{request_ref, Request};
```
Something is odd here, how does this doctest pass when it uses `core::error` without the `error_in_core` feature gate?

RalfJung replied:
E.g. if I press the "run" button [here](https://doc.rust-lang.org/nightly/std/error/trait.Error.html#method.provide), and then try to run the code in playground, I get an error.

https://github.com/rust-lang/rust/pull/114837 should fix that. But it is concerning that CI didn't catch this.
